gempy/core/tensor/tensorflow_graph.py

```diff
@@ -0,0 +1,580 @@
+
+        sub_y = tf.concat([tf.concat([tf.zeros([self.dips_position_all.shape[0], self.dips_position_all.shape[0]]), tf.ones(
+            [self.dips_position_all.shape[0], 1*self.dips_position_all.shape[0]])], axis=1), tf.zeros([self.dips_position_all.shape[0], self.dips_position_all.shape[0]])], 1)
+        sub_z = tf.concat([tf.zeros([self.dips_position_all.shape[0], 2*self.dips_position_all.shape[0]]),
+                           tf.ones([self.dips_position_all.shape[0], self.dips_position_all.shape[0]])], axis=1)
+
+        perpendicularity_matrix = tf.cast(
+            tf.concat([sub_x, sub_y, sub_z], axis=0), dtype=tf.float64)
+
+        condistion_fail = tf.math.divide_no_nan(h_u * h_v, sed_dips_dips ** 2)*(
+            tf.where(sed_dips_dips < self.a_T,
+                     x=(((-self.c_o_T * ((-14. / self.a_T ** 2.) + 105. / 4. * sed_dips_dips / self.a_T ** 3. -
+                                         35. / 2. * sed_dips_dips ** 3. / self.a_T ** 5. +
+                                         21. / 4. * sed_dips_dips ** 5. / self.a_T ** 7.))) +
+                        self.c_o_T * 7. * (9. * sed_dips_dips ** 5. - 20. * self.a_T ** 2. * sed_dips_dips ** 3. +
+                                          15. * self.a_T ** 4. * sed_dips_dips - 4.* self.a_T ** 5.) / (2. * self.a_T ** 7.)), y=0.)) -\
+            perpendicularity_matrix*tf.where(sed_dips_dips < self.a_T, x=self.c_o_T * ((-14. / self.a_T ** 2.) + 105. / 4. * sed_dips_dips / self.a_T ** 3. -
+                                                                                       35. / 2. * sed_dips_dips ** 3. / self.a_T ** 5. +
+                                                                                       21. / 4. * sed_dips_dips ** 5. / self.a_T ** 7.), y=0.)
+
+        C_G = tf.where(sed_dips_dips == 0, x=tf.constant(
+            0., dtype=self.dtype), y=condistion_fail)
+        C_G = C_G + tf.eye(C_G.shape[0],
+                           dtype=self.dtype)*self.nugget_effect_grad
+
+        return C_G
+
+    def cartesian_dist(self, x_1, x_2):
+        return tf.concat([
+            tf.transpose(
+                (x_1[:, 0] - tf.reshape(x_2[:, 0], [x_2.shape[0], 1]))),
+            tf.transpose(
+                (x_1[:, 1] - tf.reshape(x_2[:, 1], [x_2.shape[0], 1]))),
+            tf.transpose((x_1[:, 2] - tf.reshape(x_2[:, 2], [x_2.shape[0], 1])))], axis=0)
+
+    def cov_interface_gradients(self):
+
+        sed_dips_rest = self.squared_euclidean_distance(
+            self.dips_position_all_tiled, self.rest_layer_points)
+        sed_dips_ref = self.squared_euclidean_distance(
+            self.dips_position_all_tiled, self.ref_layer_points)
+
+        hu_rest = self.cartesian_dist(
+            self.dips_position_all, self.rest_layer_points)
+        hu_ref = self.cartesian_dist(
+            self.dips_position_all, self.ref_layer_points)
+
+        C_GI = self.gi_reescale*tf.transpose(hu_rest *
```

Comment:
for readability, I would recommend separating out the interiors of the tf.where calls to lines above

Reply:
Thanks for the advice Andrew, definitely, I will try to clean and polish afterwards